lsp-mode.el

```diff
@@ -5203,18 +5240,52 @@ It will show up only if current point has signature help."
         :context `( :diagnostics ,(lsp-cur-line-diagnostics)
                     ,@(when kind (list :only (vector kind))))))
 
-(defun lsp-code-actions-at-point (&optional kind)
-  "Retrieve the code actions for the active region or the current line."
-  (lsp-request "textDocument/codeAction" (lsp--text-document-code-action-params kind)))
+(defun lsp-code-actions-at-point (&optional kind enabled)
+  "Retrieve the code actions for the active region or the current line.
+If ENABLED is specified, only return code actions that are not
+disabled."
+  (--doto (lsp-request "textDocument/codeAction" (lsp--text-document-code-action-params kind))
+    (when enabled
+      (cl-callf2 seq-remove #'lsp:code-action-disabled? it))))
 
 (defun lsp-execute-code-action-by-kind (command-kind)
   "Execute code action by name."
-  (if-let ((action (->> (lsp-get-or-calculate-code-actions command-kind)
-                        (-filter (-lambda ((&CodeAction :kind?))
-                                   (and kind? (equal command-kind kind?))))
-                        lsp--select-action)))
-      (lsp-execute-code-action action)
-    (signal 'lsp-no-code-actions '(command-kind))))
+  (let* ((all-actions (->> (lsp-get-or-calculate-code-actions command-kind)
+                           (-filter (-lambda ((&CodeAction :kind?))
+                                      (and kind? (equal command-kind kind?))))))
+         (enabled-actions (-remove #'lsp:code-action-disabled? all-actions))
+         (action (lsp--select-action (or enabled-actions all-actions))))
```

Comment:
Is this ok? In case of all actions are disabled, we show all actions instead of empty list?
Also, shouldn't we need guard of `action` as the old code, in case of user not select anything.

Reply:
The idea is to show an error message in case there is no enabled code action, if the user has enabled `lsp-auto-execute-action` as t ([discussion](https://github.com/emacs-lsp/lsp-mode/pull/2402#discussion_r541268228)). However, I agree that this might be suboptimal. Do you have an idea for a better UI in this case?

As for guarding, we needn't do that, since the user cannot exit from `lsp--select-action` without either selecting an action or `quit`ting (causing a non-local exit), due to REQUIRE-MATCH being set as t.

Reply:
Isn't `lsp-auto-execute-action` is only for single code action?
The case of `lsp-execute-code-action-by-kind` is falling into the bellow in the spec, should we just take the first code action and let-bind `lsp-auto-execute-action` to `t` so the error will always be shown?
```
	 * - If the user has a keybinding that auto applies a code action and only
	 *   a disabled code actions are returned, the client should show the user
	 *   an error message with `reason` in the editor.
```

Reply:
We could `let`-bind `lsp-auto-execute-action` to t if there are no enabled ones, however I am unsure if that is the right thing: `lsp-execute-code-action-by-kind` doesn't auto-execute a single code action if available by default, either. If there is only one disabled code action, a prompt containing only it will show up, and along the code action title the `disabled-reason` will be shown:
![20210102-153855](https://user-images.githubusercontent.com/69264154/103459557-afd94780-4d10-11eb-8433-57a847251bf8.png)

Reply:
Shouldn't `lsp--select-action` automatically choose the only one code action available when `lsp-auto-execute-action` is `t`?
Is there special case for disable code action that make the `lsp--select-action` behaves otherwise?